File: app/src/main/java/com/marknkamau/justjava/ui/cart/CartActivity.java
package com.marknkamau.justjava.ui.cart;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.marknkamau.justjava.JustJavaApp;
import com.marknkamau.justjava.data.PreferencesRepository;
import com.marknkamau.justjava.ui.about.AboutActivity;
import com.marknkamau.justjava.ui.checkout.CheckoutActivity;
import com.marknkamau.justjava.R;
import com.marknkamau.justjava.ui.login.LogInActivity;
import com.marknkamau.justjava.ui.profile.ProfileActivity;
import com.marknkamau.justjava.models.CartItem;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class CartActivity extends AppCompatActivity implements FirebaseAuth.AuthStateListener, CartView {
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.btn_clear_cart)
    Button btnClearCart;
    @BindView(R.id.rv_cart)
    RecyclerView recyclerView;
    @BindView(R.id.tv_cart_total)
    TextView tvCartTotal;
    @BindView(R.id.btn_checkout)
    Button btnCheckout;
    @BindView(R.id.tv_no_items)
    TextView tvNoItems;

    private FirebaseAuth firebaseAuth;
    private FirebaseUser user;
    private CartPresenter presenter;

    @Inject
    PreferencesRepository preferencesRepository;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cart);
        ButterKnife.bind(this);

        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);

        firebaseAuth = FirebaseAuth.getInstance();

        recyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));

        ((JustJavaApp) getApplication()).getAppComponent().inject(this);
        presenter = new CartPresenter(this, preferencesRepository);
        presenter.loadItems();
    }

    @Override
    protected void onResume() {
        super.onResume();
        invalidateOptionsMenu();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        user = firebaseAuth.getCurrentUser();
        MenuInflater inflater = getMenuInflater();
        if (user != null) {
            inflater.inflate(R.menu.toolbar_menu_logged_in, menu);
        } else {
            inflater.inflate(R.menu.toolbar_menu, menu);
        }
        return true;
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_log_in:
                startActivity(new Intent(this, LogInActivity.class));
                return true;
            case R.id.menu_log_out:
                presenter.logUserOut();
                invalidateOptionsMenu();
                return true;
            case R.id.menu_profile:
                startActivity(new Intent(this, ProfileActivity.class));
                return true;
            case R.id.menu_about:
                startActivity(new Intent(this, AboutActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
        user = FirebaseAuth.getInstance().getCurrentUser();
        invalidateOptionsMenu();
    }

    @OnClick({R.id.btn_clear_cart, R.id.btn_checkout})
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btn_clear_cart:
                presenter.clearCart();
                break;
            case R.id.btn_checkout:
                startActivity(new Intent(CartActivity.this, CheckoutActivity.class));
                break;
        }
    }

    @Override
    public void displayCart(List<CartItem> cartItems) {
        CartAdapter adapter = new CartAdapter(this, cartItems, new CartAdapter.CartAdapterListener() {
            @Override
            public void updateList() {
                presenter.loadItems();
            }
        });

        recyclerView.setAdapter(adapter);
        btnCheckout.setEnabled(true);
    }

    @Override
    public void displayCartTotal(int totalCost) {
        tvCartTotal.setText(getString(R.string.total) + ": " + getString(R.string.ksh) + totalCost);
    }

    @Override
    public void displayEmptyCart() {
        tvNoItems.setVisibility(View.VISIBLE);
        btnClearCart.setEnabled(false);
        btnClearCart.setAlpha(.54f);
        tvCartTotal.setAlpha(.54f);
        tvCartTotal.setText(getString(R.string.total) + ": " + getString(R.string.ksh));
        btnCheckout.setBackgroundResource(R.drawable.large_button_disabled);
        btnCheckout.setEnabled(false);
    }



}

File: app/src/main/java/com/marknkamau/justjava/ui/cart/CartActivity.kt
package com.marknkamau.justjava.ui.cart

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.marknjunge.core.data.model.Resource
import com.marknkamau.justjava.R
import com.marknkamau.justjava.data.models.CartItem
import com.marknkamau.justjava.ui.ToolbarActivity
import com.marknkamau.justjava.ui.checkout.CheckoutActivity
import com.marknkamau.justjava.ui.login.SignInActivity
import com.marknkamau.justjava.utils.BaseRecyclerViewAdapter
import com.marknkamau.justjava.utils.CurrencyFormatter
import kotlinx.android.synthetic.main.activity_cart.*
import kotlinx.android.synthetic.main.item_cart_item.view.*
import org.koin.androidx.viewmodel.ext.android.viewModel

class CartActivity : ToolbarActivity() {

    private val cartViewModel: CartViewModel by viewModel()
    private lateinit var items: List<CartItem>

    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_cart)
        supportActionBar?.title = getString(R.string.cart)

        observeLoading()
        initializeRecyclerView()

        cartViewModel.getCartItems()

        btnCheckout.setOnClickListener {
            val clz = if (cartViewModel.isSignedIn()) CheckoutActivity::class.java else SignInActivity::class.java
            startActivity(Intent(this, clz))
        }
    }

    private fun observeLoading() {
        cartViewModel.loading.observe(this, Observer { loading ->
            pbLoading.visibility = if (loading) View.VISIBLE else View.GONE
        })
    }

    @SuppressLint("SetTextI18n")
    private fun initializeRecyclerView() {
        val adapter = BaseRecyclerViewAdapter<CartItem>(R.layout.item_cart_item) { item ->
            tv_cartItem_productName.text = item.cartItem.productName
            tv_cartItem_quantity.text = "${item.cartItem.quantity}x"

            tv_cartItem_totalPrice.text =
                getString(R.string.price_listing, CurrencyFormatter.format(item.cartItem.totalPrice))
            tv_cartItem_options.text = item.options.joinToString(", ") { it.optionName }

            rootCartItem.setOnLongClickListener {
                cartViewModel.deleteItem(item)
                true
            }
        }

        rvCart.apply {
            addItemDecoration(DividerItemDecoration(this@CartActivity, RecyclerView.VERTICAL))
            layoutManager = LinearLayoutManager(this@CartActivity, RecyclerView.VERTICAL, false)
            this.adapter = adapter
        }

        handleRecyclerViewSwipe()

        cartViewModel.items.observe(this, Observer { items ->
            this.items = items
            if (items.isEmpty()) {
                layoutCartContent.visibility = View.GONE
                layoutCartEmpty.visibility = View.VISIBLE
            } else {
                layoutCartContent.visibility = View.VISIBLE
                layoutCartEmpty.visibility = View.GONE

                adapter.setItems(items)
                val cartTotal = items.map { it.cartItem.totalPrice }.foldRight(0.0, { total, acc -> total + acc })
                tvCartTotal.text = getString(R.string.price_listing, CurrencyFormatter.format(cartTotal))

                verifyCartItems()
            }
        })
    }

    private fun handleRecyclerViewSwipe() {
        val swipeBackground = ColorDrawable(Color.parseColor("#DFDFDF"))
        val deleteIcon = ContextCompat.getDrawable(this, R.drawable.ic_delete)!!

        val touchHelperCallback =
            object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
                override fun onMove(
                    recyclerView: RecyclerView,
                    viewHolder: RecyclerView.ViewHolder,
                    target: RecyclerView.ViewHolder
                ): Boolean {
                    // Not used in swipe
                    return false
                }

                override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                    val item = items[viewHolder.adapterPosition]
                    cartViewModel.deleteItem(item)
                }

                override fun onChildDraw(
                    c: Canvas,
                    recyclerView: RecyclerView,
                    viewHolder: RecyclerView.ViewHolder,
                    dX: Float, // How far along the x-axis it has moved. > 0 = Right, < 0 = Left
                    dY: Float,
                    actionState: Int,
                    isCurrentlyActive: Boolean
                ) {
                    val itemView = viewHolder.itemView
                    val iconMargin = (itemView.height - deleteIcon.intrinsicHeight) / 2

                    if (dX > 0) {
                        swipeBackground.setBounds(itemView.left, itemView.top, dX.toInt(), itemView.bottom)
                        deleteIcon.setBounds(
                            itemView.left + iconMargin,
                            itemView.top + iconMargin,
                            itemView.left + iconMargin + deleteIcon.intrinsicWidth,
                            itemView.bottom - iconMargin
                        )
                    } else {
                        swipeBackground.setBounds(
                            itemView.right + dX.toInt(),
                            itemView.top,
                            itemView.right,
                            itemView.bottom
                        )
                        deleteIcon.setBounds(
                            itemView.right - iconMargin - deleteIcon.intrinsicWidth,
                            itemView.top + iconMargin,
                            itemView.right - iconMargin,
                            itemView.bottom - iconMargin
                        )
                    }
                    swipeBackground.draw(c)

                    // Prevent permanent changes to the canvas.
                    // Restored after clipping.
                    c.save()

                    // Prevent icon from drawing over the item by clipping it at the same points as the background color
                    if (dX > 0) {
                        c.clipRect(itemView.left, itemView.top, dX.toInt(), itemView.bottom)
                    } else {
                        c.clipRect(
                            itemView.right + dX.toInt(),
                            itemView.top,
                            itemView.right,
                            itemView.bottom
                        )
                    }
                    deleteIcon.draw(c)

                    // Restore the state of the canvas before clipping to prevent clipping elements that shouldn't
                    c.restore()

                    super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
                }
            }

        val itemTouchHelper = ItemTouchHelper(touchHelperCallback)
        itemTouchHelper.attachToRecyclerView(rvCart)
    }

    private fun verifyCartItems() {
        cartViewModel.verifyOrder(items).observe(this, Observer { resource ->
            btnCheckout.isEnabled = true
            when (resource) {
                is Resource.Success -> {
                    if (resource.data.isNotEmpty()) {
                        // TODO Handle changed products
                    }
                }
                is Resource.Failure -> handleApiError(resource)
            }
        })
    }
}
